File: PASCAL/Cap07/c07ex09.pas
program c07ex09;

function FATORIAL_BASE(N, P: Integer): Integer;
begin
  if (N = 0) then
    FATORIAL_BASE := P
  else
    FATORIAL_BASE := FATORIAL_BASE(N - 1, N * P);
end;

function FATORIAL(N: Integer): Integer;
begin
  FATORIAL := FATORIAL_BASE(N, 1);
end;

var
  LIMITE: Integer;

begin
  Write('Qual fatorial: ');
  Readln(LIMITE);

  Writeln('Fatorial = ', FATORIAL(LIMITE));

  Writeln;
  Write('Tecle <Enter> para encerrar... ');
  Readln;
end.

File: C/Cap07/c07ex09.c
// c07ex09.c

#include <stdio.h>

int FATORIAL_BASE(int N, int P) {
  if (N == 0)
    return P;
  else
    return FATORIAL_BASE(N - 1, N * P);
}

int FATORIAL(int N) {
  return FATORIAL_BASE(N, 1);
}

int main(void) {
  int LIMITE;

  printf("Qual fatorial: ");
  scanf("%d", &LIMITE);
  while (getchar() != '\n');

  printf("Fatorial = %d\n", FATORIAL(LIMITE));

  printf("\n");
  printf("Tecle <Enter> para encerrar... ");
  while (getchar() != '\n');

  return 0;
}
